tests/aws/services/lambda_/test_lambda_performance.py

```diff
@@ -81,6 +103,520 @@ def invoke():
+            error_counter.increment()
+
+    start_time = datetime.utcnow()
+    # Use threads to invoke Lambda function in parallel
+    thread_list = []
+    for i in range(1, num_invocations + 1):
+        thread = threading.Thread(target=invoke, args=[i])
+        thread.start()
+        thread_list.append(thread)
+
+    for thread in thread_list:
+        thread.join()
+    end_time = datetime.utcnow()
+    diff = end_time - start_time
+    LOG.info(f"N={num_invocations} took {diff.total_seconds()} seconds")
+    assert error_counter.counter == 0
+
+    # Sleeping here is a bit hacky, but we want to avoid polling for now because polling affects the results.
+    sleep_seconds = 2000
+    LOG.info(f"Sleeping for {sleep_seconds} ...")
+    time.sleep(sleep_seconds)
+
+    # Validate CloudWatch invocation metric
+    def assert_cloudwatch_metric():
+        metric_query_params = {
+            "Namespace": "AWS/Lambda",
+            "MetricName": "Invocations",
+            "Dimensions": [{"Name": "FunctionName", "Value": function_name}],
+            "StartTime": start_time,
+            # TODO: are potentially future times valid?
+            "EndTime": end_time + timedelta(seconds=10),
+            "Period": 3600,  # in seconds
+            "Statistics": ["Sum"],
+        }
+        response = aws_client.cloudwatch.get_metric_statistics(**metric_query_params)
+        num_invocations_metric = 0
+        for datapoint in response["Datapoints"]:
+            num_invocations_metric += datapoint["Sum"]
+        # assert num_invocations_metric == num_invocations
+        return num_invocations_metric
+
+    metric_count = assert_cloudwatch_metric()
+    # retry(assert_cloudwatch_metric, retries=300, sleep=10)
+
+    # Validate CloudWatch invocation logs
+    def assert_log_events():
+        # the default and maximum limit is 10k events
+        response = aws_client.logs.filter_log_events(
+            logGroupName=f"/aws/lambda/{function_name}",
+        )
+        assert "nextToken" not in response  # guard against pagination
+        log_events = response["events"]
+        invocation_count = len(
+            [event["message"] for event in log_events if event["message"].startswith("REPORT")]
+        )
+        # assert invocation_count == num_invocations
+        return invocation_count
+
+    log_count = assert_log_events()
+    # NOTE: slow against AWS (can take minutes and would likely require more retries)
+    # retry(assert_log_events, retries=300, sleep=2)
+
+    # Validate S3 object creation
+    def assert_s3_objects():
+        s3_keys_output = get_s3_keys(aws_client, s3_bucket)
+        # assert len(s3_keys_output) == num_invocations
+        return len(s3_keys_output)
+
+    s3_count = assert_s3_objects()
+
+    # TODO: the CloudWatch metric count for `Invocations` seems very unreliable => double-check API and implementation
+    assert [metric_count, log_count, s3_count] == [
```

Comment:
@pinzon @steffyP For 2000 async invokes, I'm getting metrics like `[10.0, 1934, 1942]` 😬 

Hence, only `10.0` of 2k concurrent `Invocations` are tracked in CloudWatch v1. Do you have any idea why (user error/wrong query, concurrency issue)?

Reply:
FYI: concurrency is an issue (but I assume not that bad)

```
2023-12-15T17:26:53.250  INFO --- [sqs-metrics-cloudwatch-dispatcher_1] localstack.utils.cloudwatch.cloudwatch_util : Unable to put metric data for metric "NumberOfEmptyReceives" to CloudWatch: An error occurred (InternalError) when calling the PutMetricData operation (reached max retries: 0): exception while calling cloudwatch with unknown operation: Traceback (most recent call last):
  File "/Users/joe/Projects/LocalStack/localstack/localstack/aws/chain.py", line 90, in handle
    handler(self, self.context, response)
  File "/Users/joe/Projects/LocalStack/localstack/localstack/aws/handlers/routes.py", line 27, in __call__
    router_response = self.router.dispatch(context.request)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/joe/Projects/LocalStack/localstack/localstack/http/router.py", line 440, in dispatch
    handler, args = matcher.match(get_raw_path(request), method=request.method)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/joe/Projects/LocalStack/localstack-ext/.venv/lib/python3.11/site-packages/werkzeug/routing/map.py", line 582, in match
    self.map.update()
  File "/Users/joe/Projects/LocalStack/localstack-ext/.venv/lib/python3.11/site-packages/werkzeug/routing/map.py", line 365, in update
    for rules in self._rules_by_endpoint.values():
RuntimeError: dictionary changed size during iteration
```

Reply:
yes, concurrency is a known issue with the current moto fallback implementation of cloudwatch.